README.md
Fx.Tabs - An Advanced Tab Swapper
=================================

Fx.Tabs is the new name for MorphTabs.

Fx.Tabs is a complete clone of mooTabs by [silverscripting](http://www.silverscripting.com/mootabs/). I have updated it for mootools 1.2.4 and it will not work with mootool 1.11 and below.

I have also added some new effects which I hope you all will like!

Because Fx.Tabs is based heaverly on mootabs some of the options and documentation is the same, but there are some important changes to note.

Inherits methods, properties, options and events from [Tabs][].

![Screenshot](http://github.com/vincentbluff/Fx.Tabs/raw/master/screenshot.png)

###Extends
* [Tabs][]

Requirements
------------

* [MooTools Core 1.2.4](http://mootools.net/core): Element.Event, Selectors, DomReady, Fx.Morph, Fx.Transitions and their dependencies
* [fx.MorphElement][] and its dependencies
* [ElementSwap][] and its dependencies
* [Fx.ElementSwap][] and its dependencies
* [Tabs][] and its dependencies

How to use
----------

### Syntax
	new Fx.Tabs([options]);

### Arguments

1. options - (*object*, optional) An object with options. See:
	- [Tabs][]
	- [Fx.ElementSwap][]
	- [ElementSwap][]
	- [fx.MorphElement][]
for all available options.

### Example

### The JavaScript part

	#JS
	window.addEvent('domready', function() {
		var fxTabs = new Fx.Tabs();
	});

### The XHTML part

	#HTML
	<div id="tabs">

		<!-- tab headings -->

		<ul class="tabs_title">
			<li title="my_work">My Work</li>
			<li title="about_me">About Me</li>

			<li title="contact">Contact</li>
		</ul>

			
		<!-- Tab1 content -->

		<div id="my_work" class="tabs_panel">
			<h1>Tab 1</h1>
			<p>content</p>
		</div>

		<!-- Tab2 content -->

		<div id="about_me" class="tabs_panel">
			<h1>Tab 2</h1>
			<p>content</p>
		</div>
			
		<!-- Tab3 content -->

		<div id="contact" class="tabs_panel">
			<h1>Tab 3</h1>
			<p>content</p>
		</div>
	
	</div>

###The CSS part

Customizing the look of the tabs require a certain knowledge of CSS. What I would suggest is to look at the examples of tabs in the the Demo folder, they should be a good starting point.

#### Example

	#CSS
	#tabs {
	position:relative;
	width:300px;
	height:200px;
	overflow:hidden;
	}
	
	.tabs_title {
		list-style-image: none;
		list-style-type: none;
		margin: 0px;
		padding: 0px;
		height: 24px;
	}
	
	.tabs_title li {
		float: left;
		background-color: #3975BD;
		padding: 2px 8px 2px 8px;
		margin-right: 2px;
		cursor: pointer;
		color: #fff;
		font-family: "Trebuchet MS";
		font-size: 12px;
		height: 24px;
		line-height: 24px;
	}
	
	.tabs_title a {
		text-decoration:none;
		color: #fff;
	}
	
	.tabs_title li.over {
		font-weight: bold;
	}
	
	.tabs_title li.active {
		background-color: #49A8EC;
	}
	
	.tabs_panelwrap {
		position: absolute;
		top:28px;
		overflow: hidden;
		width:300px;
		height:172px;
	}
	
	.tabs_panel {
		/*position:absolute;
		top:28px;*/
		width:300px;
		height:172px;
		display:none;
		overflow: auto;
		background-color: #49A8EC;
		color: #fff;
		clear: both;
	}
	
	.tabs_panel.active {
		display: block;
	}
	
	.tabs_panel p {
		margin-left:5px;
	}

The styling is completly left to you this is why I would suggest you to read up a bit on CSS if you are new to it, before you start styling the tabs.

[$$]: http://www.mootools.net/docs/core/Element/Element#dollars
[ElementSwap]: http://www.mootools.net/forge/p/elementswap
[Fx.ElementSwap]: http://www.mootools.net/forge/p/fx_elementswap
[Tabs]: http://www.mootools.net/forge/p/tabs
[Fx.MorphElement]: http://www.mootools.net/forge/p/fx_morphelement

CHANGES
version 1.0 - Mon June 9 2008
	- 1st release
version 1.1 - Mon Sep 22 2008
	- New Options:
		- added activateTabFunction
		- added evalScripts
version 1.2 - Wed Oct 8 2008
	- New Methods:
		- added start
		- added stop
	- New Options:
		- added slideShow
		- added slideShowDelay
version 1.3 - Web Nov 19 2008
	- bug fix:
		- line 87 : issue with IE and NaN fix by Rob Verzera
version 1.4 - Fri Dec 12 2008
	- bug fix:
		- line 59 : issue with lists in panel fixed.
version 1.5 - Wed Dec 17 2009
	- changed name to Fx.Tabs
version 2.0 - Fri Dec 25 2009
	- complete rewrite

Demo/js/Tabs.js
/*
---
name: Tabs.js
description: Handles the scripting for a common UI layout; the tabbed box. 
authors: Shaun Freeman
requires:
    core/1.2.4
    - Selectors
    - DomReady
    - HTML.Request
    more/1.2.4
    - URI
    elementswap/1.0
    - ElementSwap
provides: [Tabs]
license: MIT-style license
version: 1.0
...
*/

var Tabs = new Class({
	Implements: [Options, Events],
 
	options: {
		tabs: '.tabs_title li',
		panels: '.tabs_panel',
		selectedClass: 'active',
		elementSwapOptions: {
			selectedClass: 'active',
			panelWrap: 'tabsPanelWrap',
			panelWrapClass: 'tabs_panelwrap',
			activateOnLoad: 0,
			events: true,
			autoPlay: false
		},
		mouseOverClass: 'over',
		onActive: $empty,
		onBackground: $empty
	},
		
	initialize: function(options) {
		this.setOptions(options);
		this.tabs = $$(this.options.tabs);
		
		this.getSwap();
		
		this.attach(this.tabs);
		
		if($type(this.options.elementSwapOptions.activateOnLoad) == 'number') {
			this.activateTab(this.options.elementSwapOptions.activateOnLoad);
		}
	},
	
	getSwap: function(){
		this.elSwap = new ElementSwap(this.options.panels, this.options.elementSwapOptions).addEvent('onActive', this.showTab.bindWithEvent(this));
	},
	
	attach: function(elements) {
		$$(elements).each(function(element) {
			var enter = element.retrieve('tab:enter', this.elementEnter.bindWithEvent(this, element));
			var leave = element.retrieve('tab:leave', this.elementLeave.bindWithEvent(this, element));
			var mouseclick = element.retrieve('tab:click', this.elementClick.bindWithEvent(this, element));
			element.addEvents({
				mouseenter: enter,
				mouseleave: leave,
				click: mouseclick
			});
		}, this);
		return this;
	},
	
	activateTab: function(index) {
		this.showTab(index);
		this.elSwap.activate(index);
	},
	
	showTab: function(index) {
		this.now = index;
		this.tabs.removeClass(this.options.selectedClass);
		this.tabs[this.elSwap.now].addClass(this.options.selectedClass);
		this.fireEvent('onActive', [this.tabs[this.now], index]);
		return this;
	},
	
	elementEnter: function(event, element) {
		if(element != this.tabs[this.now]) {
			element.addClass(this.options.mouseOverClass);
		}
	},
 
	elementLeave: function(event, element) {
		if(element != this.tabs[this.now]) {
			element.removeClass(this.options.mouseOverClass);
		}
	},
	
	elementClick: function(event, element) {
		event.stop();
		if(element != this.tabs[this.now]) {
			if (this.elSwap.isLooping) this.elSwap.stop(true);
			this.fireEvent('onBackground', [this.tabs[this.now]]);
			this.activateTab(this.tabs.indexOf(element));
		}
	}
});


Demo/js/selectTrans.js
var selectTrans = new Class({
	
	Implements: [Options],
	
	options: {
		transition: [
			'select',
			'linear',
			'quad:in', 'quad:out', 'quad:in:out',
			'cubic:in', 'cubic:out', 'cubic:in:out',
			'quart:in', 'quart:out', 'quart:in:out',
			'quint:in', 'quint:out', 'quint:in:out',
			'expo:in', 'expo:out', 'expo:in:out',
			'circ:in', 'circ:out', 'circ:in:out',
			'sine:in', 'sine:out', 'sine:in:out',
			'back:in', 'back:out', 'back:in:out',
			'bounce:in', 'bounce:out', 'bounce:in:out',
			'elastic:in', 'elastic:out', 'elastic:in:out'
		 ],
		 duration: [
			'select',
			'short', 'normal', 'long'
		],
		commonFx: [
			'select',
			'blind:up', 'blind:down', 'blind:left', 'blind:right',
			'slide:up', 'slide:down', 'slide:left', 'slide:right'
		],
		panelStartFx: ['fade'],
		panelEndFx: ['appear']
	},
 
	initialize: function(element,morphTabs) {
		
		var divs = $$('#' + element + ' div');
		
		divs.each(function(element) {
			
			var selectOptions = null;
			
			var selectBox = new Element('select').inject(element);
			
			var opt = element.id.split("_");
			
			if (opt[0].contains('Fx')) {
				this.fx = [];
				
				this.fx = this.options.commonFx;
				if (opt[0].contains('Start')) {
					selectOptions = this.fx.include(this.options.panelStartFx);
				} else {
					selectOptions = this.fx.include(this.options.panelEndFx).erase(this.options.panelStartFx);
				}
			} else {
				selectOptions = this.options[opt[opt.length - 1]];
			}
			
			selectOptions.each(function(selectValue) {
				new Element('option', {
					'value': selectValue,
					'text': selectValue
				}).inject(selectBox);
			});
			
			var selectChange = selectBox.retrieve('selectBox:change', this.elementChange.bindWithEvent(this, morphTabs));
			
			selectBox.addEvents({
				change: selectChange
			});
		},this);
	},
 
	elementChange: function(event, morphTabs) {
		var event = new Event(event).stop();
		var changeValue = event.target.get('value');
		var opt = event.target.getParent().id.split("_");
		if (opt.length == 1) {
			morphTabs.options[opt[0]] = changeValue;
		} else {
			morphTabs.options[opt[0]][opt[1]] = changeValue;
		}
	}
});

Demo/js/Fx.ElementSwap.js
/*
---
name: Fx.ElementSwap.js
description: Slide show interface for classes.
authors: Shaun Freeman
requires:
    fx_morphelement/1.0.3:
    - Fx.MorphElement
    - Fx.MorphElement.Effects
    elementswap/1.0.1:
    - ElementSwap
provides: [Fx.ElementSwap]
license: MIT-style license
version: 1.0.2
...
*/

Fx.ElementSwap = new Class ({
	
	Extends: ElementSwap,
 
	options: {
		TransitionFx: {
			transition: 'linear',
			duration: 'long',
			onStart: function() {
				this.element.setStyle('overflow', 'hidden');
			},
			onComplete: function() {
				this.element.setStyle('overflow', 'auto');
			}
		},
		showFx: 'slide:right',
		hideFx: 'slide:left',
		wait: true
	},
	
	firstRun: true,
	
	initialize: function(elements, options) {
		this.setOptions(options);
		
		this.slides = $$(elements);
		
		this.wrap = new Element('div', {
			'id': this.options.panelWrap,
			'class': this.options.panelWrapClass
		}).inject(this.slides[0], 'before').adopt(this.slides);
		
		this.attachFx(this.slides);
		
		this.activate(this.options.activateOnLoad);
		
		if (this.options.autoPlay) this.start();
	},
	
	attachFx: function(elements) {
		$$(elements).each(function(el) {
			el.get('morphElement', {
				wrap: false,
				width: document.id(this.options.panelWrap).getStyle('width'),
				height: document.id(this.options.panelWrap).getStyle('height'),
				FxTransition: this.options.TransitionFx,
				hideOnInitialize: this.options.hideFx
			});
		}, this);
		return this;
	},
	
	activate: function(index) {
		if ($type(index) == 'string') index = this.slides.indexOf(this.slides.filter('[id='+index+']')[0]);
		if ($type(index) != 'number') return;
		// panel fx here..
		if (this.firstRun) {
			this.show(index);
			this.firstRun = false;
		} else {
			if (this.options.wait) {
				this.getFx('hideFx').chain(
					function() {
						this.show(index);
					}.bind(this)
				);
			} else {
				this.getFx('hideFx');
				this.show(index);
			}
		}
	},
	
	show: function(index) {
		this.parent(index);
		this.getFx('showFx');
	},
	
	getFx: function(fx) {
		this.effect = this.slides[this.now].retrieve('morphElement');
		return this.effect.start(this.options[fx]);
	},
	
	changeFx: function(elements, fx) {
		if (elements == 'all') elements = this.slides;
		fx = {
			FxTransition: fx,
			hideOnInitialize: false
		};
		$$(elements).each(function(el) {
			var flag = el.retrieve('fxEffect:flag');
			opts = el.retrieve('morphElement:options');
			el.eliminate('fxEffect:flag').eliminate('morphElement');
			el.get('morphElement', $merge(opts, fx));
		}, this);
		this.slides[this.now].store('fxEffect:flag', 'hide');
	}
	
});

Element.implement({

	fxSwap: function(options) {
		return new Fx.ElementSwap(this.getChildren(),options);
	}

});

Demo/js/Fx.MorphElement.js
/*
---
name: Fx.MorphElement.js
description: Creates a morph effect on any element using predefined styles.
authors: Shaun Freeman
requires:
    core:1.2.4:
    - Fx.Morph
    - Fx.Transitions
provides: [Fx.MorphElement, Fx.MorphElement.Effects]
license: MIT-style license
version: 1.0.2
...
*/

Fx.MorphElement = new Class({
	
	Extends: Fx.Morph,
	
	options: {
		wrap: true,
		wrapClass: 'morphElementWrap',
		FxTransition : $empty,
		hideOnInitialize: false,
		width: null,
		height: null
	},
	
	initialize: function(el, options) {
		this.setOptions(options);
		this.parent(el, this.options.FxTransition);
		
		if (!this.options.width) this.setOptions({width: this.element.getWidth()});
		if (!this.options.height) this.setOptions({height: this.element.getHeight()});
		
		if (this.options.wrap) {
			this.wrap = new Element('div', {
				'id': this.element.get('id') + '_wrap',
				'class': this.options.wrapClass,
				'styles': {
					'height': this.options.height,
					'width': this.options.width,
					'overflow': 'hidden'
				}
			}).wraps(this.element);
		}
		
		this.element.setStyle('overflow', 'auto');
		
		this.element.store('fxEffect:flag', 'hide');
		
		if (this.options.hideOnInitialize) {
			this.start(this.options.hideOnInitialize);
		} else {
			this.element.store('fxEffect:flag', 'show');
			this.set({'opacity': [1,1]});
		}
	},
	
	start: function(fx) {
		
		var flag = this.element.retrieve('fxEffect:flag');
		
		var styles = {
			'margin-top': [0, 0],
			'margin-left': [0, 0],
			'width': [this.options.width, this.options.width],
			'height': [this.options.height, this.options.height],
			'opacity': [1, 1]
		};
		
		fx = fx.split(':');
		
		if (fx.length > 1) {
			fx = Fx.MorphElement.Effects[fx[0]][fx[1]][flag];
		} else {
			fx = Fx.MorphElement.Effects[fx[0]][flag];
		}
		
		$H(fx).each(function(hash, hashIndex){
			hash.each(function(item, index){
				if ($type(item) == 'string') {
					hash[index] = item.substitute({'width': this.options.width, 'height': this.options.height});
				}
			}.bind(this));
			styles[hashIndex] = hash
		}.bind(this));
		
		this.element.store('fxEffect:flag', (flag == 'hide') ? 'show' : 'hide');
		
		return this.parent(styles);
	}
});

Element.Properties.morphElement = {

	set: function(options){
		var morphElement = this.retrieve('morphElement');
		if (morphElement) morphElement.cancel();
		return this.eliminate('morphElement').store('morphElement:options', $extend({link: 'cancel'}, options));
	},

	get: function(options){
		if (options || !this.retrieve('morphElement')){
			if (options || !this.retrieve('morphElement:options')) this.set('morphElement', options);
			this.store('morphElement', new Fx.MorphElement(this, this.retrieve('morphElement:options')));
		}
		return this.retrieve('morphElement');
	}

};

Element.implement({

	morphElement: function(props){
		this.get('morphElement').start(props);
		return this;
	}

});

Fx.MorphElement.Effects = $H({
	blind: {
		up: {
			hide: {'height': ['{height}', 0]},
			show: {
				'margin-top': ['{height}', 0],
				'height': [0, '{height}']
			}
		},
		down: {
			hide: {
				'margin-top': ['{height}'],
				'height': [0]
			},
			show: {'height': [0, '{height}']}
		},
		left: {
			hide: {'width': ['{width}', 0]},
			show: {
				'margin-left': ['{width}', 0],
				'width': [0, '{width}']
			}
		},
		right: {
			hide: {
				'margin-left': ['{width}'],
				'width': [0]
			},
			show: {'width': [0, '{width}']}
		}
	},
	slide: {
		up: {
			hide: {
				'margin-top': [0, '-{height}'],
				'width': ['{width}'],
				'height': ['{height}']
			},
			show: {'margin-top': ['{height}', 0]}
		},
		down: {
			hide: {
				'margin-top': [0, '{height}'],
				'width': ['{width}'],
				'height': ['{height}']
			},
			show: {'margin-top': ['-{height}', 0]}
		},
		left: {
			hide: {
				'margin-left': [0, '-{width}'],
				'width': ['{width}'],
				'height': ['{height}']
			},
			show: {'margin-left': ['{width}', 0]}
		},
		right: {
			hide: {
				'margin-left': [0, '{width}'],
				'width': ['{width}'],
				'height': ['{height}']
			},
			show: {'margin-left': ['-{width}', 0]}
		}
	},
	fade: {
		hide: {'opacity': [1, 0]},
		show: {'opacity': [0, 1]}
	}
});


Source/MorphTabs.js
/*
---

name: MorphTabs.js
description: Handles the scripting for a common UI layout; the tabbed box.
authors: 
	- Shaun Freeman
license: 
	- MIT-style
requires:
	core/1.2: [Selectors, DomReady, Fx.Morph, Fx.Transitions, Request.HTML]
provides:
	- MorphTabs
version: 1.4

...
*/

var MorphTabs = new Class({
	Implements: [Options, Chain],
 
	version: '1.4',
 
	options: {
		width: '300px',
		height: '200px',
		changeTransition: {
			transition: 'linear',
			duration: 'long'
		},
		panelStartFx: 'blind:left',
		panelEndFx: 'blind:right',
		mouseOverClass: 'over',
		activateOnLoad: 'first',
  		activateTabFunction: $empty,
		evalScripts: false,
		useAjax: false,
		ajaxUrl: '',
		ajaxOptions: {},
		ajaxLoadingText: 'Loading...',
		slideShow: false,
		slideShowDelay: 3
	},
 
	initialize: function(element, options) {
		
		this.setOptions(options);
		this.el = $(element);
		this.elid = element;
		
		this.el.setStyles({
			'height': this.options.height,
			'width': this.options.width
		});
		 
		this.titles = $$('#' + this.elid + ' ul.morphtabs_title li');
		//console.log(this.titles);
		
		this.panelHeight = this.options.height.toInt() - (this.titles[0].getSize().y + 4);
		this.panelWidth = this.el.getSize().x;
			
		//this.el.getSize().y
		this.panel = new Element('div', {
			'id': 'morphPanel',
			'class': 'morphtabs_panel',
			'styles': {
				'width': this.panelWidth + 'px',
				'height': this.panelHeight + 'px'
			}
		}).inject(this.el.getFirst(), 'after');
		
		this.panelWrapBorder = this.panel.getStyle('border-width').toInt() * 2;
		
		this.panelWrap = new Element('div', {
			'id': 'morphPanelWrap',
			'class': 'morphtabs_panelwrap',
			'styles': {
				'height': (this.panelHeight + this.panelWrapBorder) + 'px',
				'width': (this.panelWidth + this.panelWrapBorder) + 'px'
			}
		}).wraps(this.panel);
		
		this.panelTop = this.panelWrap.getStyle('top').toInt();
		
		//if (this.panelTop == 'NaN') this.panelTop = 0;
		if (isNaN(this.panelTop)) this.panelTop = 0; // fix by Rob Verzera.
		
		this.panelWrap.setStyle('top', (Browser.Engine.trident5) ? (this.panelTop + 4) + 'px' : this.panelTop + 'px'); // fix for ie7.
		
		this.attach(this.titles);
		
		if(this.options.activateOnLoad != 'none') {
			this.firstRun = true;
			if(this.options.activateOnLoad == 'first') {
				this.activate(this.titles[0]);
			} else {
				this.activate(this.options.activateOnLoad);
			}
		}
		
		if (this.options.slideShow) this.start();
	},
 
	attach: function(elements) {
		
		$$(elements).each(function(element) {
			
			var enter = element.retrieve('tab:enter', this.elementEnter.bindWithEvent(this, element));
			
			var leave = element.retrieve('tab:leave', this.elementLeave.bindWithEvent(this, element));
			
			var mouseclick = element.retrieve('tab:click', this.elementClick.bindWithEvent(this, element));
			
			element.addEvents({
				mouseenter: enter,
				mouseleave: leave,
				click: mouseclick
			});
			
			var el = $(element.get('title'));
			element.store('panel:html', el.get('html'));
			element.store('panel:id', el.id);
			var elementDispose = $(element.get('title')).dispose();
			
		}, this);
		
		return this;
	},
 
	detach: function(elements) {
		
		$$(elements).each(function(element){
			
			element.removeEvent('mouseenter', element.retrieve('tab:enter') || $empty);
			
			element.removeEvent('mouseleave', element.retrieve('tab:leave') || $empty);
			
			element.removeEvent('mouseclick', element.retrieve('tab:click') || $empty);
			
			element.eliminate('tab:enter').eliminate('tab:leave').eliminate('tab:click').eliminate('panel:html').eliminate('panel:id');
			
			var elementDispose = element.dispose();
		});
		
		return this;
	},
 
	activate: function(tab) {
		
		if($type(tab) == 'string') {
			myTab = $$('#' + this.elid + ' ul li').filter('[title=' + tab + ']')[0];
			tab = myTab;
		}
		
		if($type(tab) == 'element') {
			
			var html = tab.retrieve('panel:html');
			this.panel.id = tab.retrieve('panel:id');
			this.titles.removeClass('active');
			tab.addClass('active');
			this.activeTitle = tab;
			this.panel.setStyle('overflow', 'hidden');
			
			if ($type(this.options.changeTransition) == 'object' && !this.firstRun) {
				
				this.getPanelFx(this.options.panelStartFx).chain(function() {
					
					this.fill(this.panel, html);
					
					this.getPanelFx(this.options.panelEndFx).chain(function() {
						
						this.options.activateTabFunction(this.panel.id);
						this.panel.setStyle('overflow', 'auto');
						this.panel.scrollTo(0,0);
						
					}.bind(this));
				}.bind(this));
				
			} else if ($type(this.options.changeTransition) == 'object' && this.firstRun) {
				
				this.fill(this.panel, html);
				
				this.panel.store('flag', 'end');
				
				this.getPanelFx(this.options.panelEndFx).chain(function() {
					
					this.options.activateTabFunction(this.panel.id);
					this.panel.setStyle('overflow', 'auto');
					this.panel.scrollTo(0,0);
					
				}.bind(this));
			} else {
				
				this.fill(this.panel, html);
				this.options.activateTabFunction(this.panel.id);
				this.panel.setStyle('overflow', 'auto');
				this.panel.scrollTo(0,0);
			}
		}
		
	},
 
	elementEnter: function(event, element) {
		
		if(element != this.activeTitle) {
			element.addClass(this.options.mouseOverClass);
		}
	},
 
	elementLeave: function(event, element) {
		
		if(element != this.activeTitle) {
			element.removeClass(this.options.mouseOverClass);
		}
	},
 
	elementClick: function(event, element) {
		
		if(element != this.activeTitle) {
			element.removeClass(this.options.mouseOverClass);
			this.activate(element);
		}
		
		if (this.slideShow) {
			this.setOptions(this.slideShow, false);
			this.clearChain();
			this.stop();
			this.panel.store('fxEffect:flag', 'show');
		}
		
	},
 
	fill: function(element, contents) {
		
		if(this.options.useAjax) {
			this.getContent();
		} else {
			
			element.set('html', contents);
			if (this.options.evalScripts) {
				element.get('html').stripScripts(true);
			}
		}
	},
 
	getContent: function() {
		
		this.panel.set('html', this.options.ajaxLoadingText);
		
		var newOptions = {
			url: this.options.ajaxUrl + '?tab=' + this.activeTitle.getProperty('title'),
			update: this.panel
		};
		
		this.options.ajaxOptions = $merge(this.options.ajaxOptions, newOptions);
		var tabRequest = new Request.HTML(this.options.ajaxOptions);
		tabRequest.send();
	},
 
	addTab: function(title, label, content) {
		
		var newTitle = new Element('li', {
			'title': title
		});
		
		newTitle.appendText(label);
		
		this.titles.include(newTitle);
		
		$$('#' + this.elid + ' ul').adopt(newTitle);
		
		var newPanel = new Element('div', {
			'id': title,
			'class': 'mootabs_panel'
		});
		
		if(!this.options.useAjax) {
			newPanel.set('html',content);
		}
		
		this.el.adopt(newPanel);
		this.attach(newTitle);
	},
 
	removeTab: function(title){
		
		if(this.activeTitle.title == title) {
			this.activate(this.titles[0]);
		}
		
		var tab = $$('#' + this.elid + ' ul li').filter('[title=' + title + ']')[0];
		this.detach(tab);
	},
 
	start: function() {
		this.slideShow = this.next.periodical(this.options.slideShowDelay * 1000, this);
	},
 
	stop: function() {
		$clear(this.slideShow);
	},
 
	next: function() {
		var nextTab = this.activeTitle.getNext();
		
		if(!nextTab) {
			nextTab = this.titles[0];
		}
		
		this.activate(nextTab);
	},
 
	previous: function() {
		var previousTab = this.activeTitle.getPrevious();
		
		if(!previousTab) {
			previousTab = this.titles[this.titles.length - 1];
		}
		
		this.activate(previousTab);
	},
	
	getPanelFx: function(fx) {
		
		this.flag = (this.firstRun) ? this.panel.retrieve('fxEffect:flag', 'show') : this.panel.retrieve('fxEffect:flag');
		
		var styles = {
			'margin-top': [0, 0],
			'margin-left': [0, 0],
			'width': [this.panelWidth, this.panelWidth],
			'height': [this.panelHeight, this.panelHeight],
			'opacity': [1, 1]
		};
		
		fxEffect = this.panel.get('morph', this.options.changeTransition);
		
		switch(fx) {
			case 'blind:up':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'height': [this.panelHeight, 0]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-top': [this.panelHeight, 0],
							'height': [0, this.panelHeight]
						}));
					}
					break;
			case 'blind:down':
					
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'margin-top': [this.panelHeight],
							'height': [0]
						}));	
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'height': [0, this.panelHeight]
						}));
					}
					break;
			case 'blind:left':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'width': [this.panelWidth, 0]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-left': [this.panelWidth, 0],
							'width': [0, this.panelWidth]
						}));
					}
					break;
			case 'blind:right':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'margin-left': [this.panelWidth],
							'width': [0]
						}));
					} else {
						styles = fxEffect.start($merge(styles, {
							'width': [0, this.panelWidth]
						}));
					}
					break;
			case 'slide:up':
					if (this.flag == 'hide') {
					
						styles = fxEffect.start($merge(styles, {
							'margin-top': [0, -this.panelHeight],
							'width': [this.panelWidth],
							'height': [this.panelHeight]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-top': [this.panelHeight, 0]
						}));
					}
					break;
			case 'slide:down':
					
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'margin-top': [0, this.panelHeight],
							'width': [this.panelWidth],
							'height': [this.panelHeight]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-top': [-this.panelHeight, 0]
						}));
					}
					break;
			case 'slide:left':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'margin-left': [0, -this.panelWidth],
							'width': [this.panelWidth],
							'height': [this.panelHeight]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-left': [this.panelWidth, 0]
						}));
					}
					break;
			case 'slide:right':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'margin-left': [0, this.panelWidth],
							'width': [this.panelWidth],
							'height': [this.panelHeight]
						}));
					} else {
						
						styles = fxEffect.start($merge(styles, {
							'margin-left': [-this.panelWidth, 0]
						}));
					}
					break;
			case 'fade':
					if (this.flag == 'hide') {
				
						styles = fxEffect.start($merge(styles, {
							'opacity': [1, 0]
						}));
					}
					break;
			case 'appear':
					if (this.flag == 'show') {
						
						styles = fxEffect.start($merge(styles, {
							'opacity': [0, 1]
						}));
					}
					break;
		}
		
		this.panel.store('fxEffect:flag', (this.flag == 'hide') ? 'show' : 'hide');
		
		if (this.firstRun) this.firstRun = false;
		
		return styles;
	}
});
